crates/ruff/src/renamer.rs

```diff
@@ -0,0 +1,176 @@
+    ///     y = 2
+    ///     print(y)
+    ///
+    /// print(y)
+    /// ```
+    ///
+    /// ## Limitations
+    ///
+    /// `global` and `nonlocal` declarations are not yet supported.
+    ///
+    /// `global` and `nonlocal` declarations add some additional complexity. If we're renaming a
+    /// name that's declared as `global` or `nonlocal` in a child scope, we need to rename the name
+    /// in that scope too, repeating the above process.
+    ///
+    /// If we're renaming a name that's declared as `global` or `nonlocal` in the current scope,
+    /// then we need to identify the scope in which the name is declared, and perform the rename
+    /// in that scope instead (which will in turn trigger the above process on the current scope).
+    pub(crate) fn rename(
+        name: &str,
+        target: &str,
+        scope: &Scope,
+        semantic: &SemanticModel,
+    ) -> Result<(Edit, Vec<Edit>)> {
+        let mut edits = vec![];
+
+        // Iterate over every binding to the name in the scope.
+        for binding_id in scope.get_all(name) {
+            let binding = semantic.binding(binding_id);
+
+            // Rename the binding.
+            if let Some(edit) = Renamer::rename_binding(binding, target) {
+                edits.push(edit);
+
+                // Rename any delayed annotations.
+                if let Some(annotations) = semantic.delayed_annotations(binding_id) {
+                    edits.extend(annotations.iter().filter_map(|annotation_id| {
+                        let annotation = semantic.binding(*annotation_id);
+                        Renamer::rename_binding(annotation, target)
+                    }));
+                }
+
+                // Rename the references to the binding.
+                edits.extend(binding.references().map(|reference_id| {
+                    let reference = semantic.reference(reference_id);
+                    Edit::range_replacement(target.to_string(), reference.range())
+                }));
+            }
+        }
+
+        // Deduplicate any edits. In some cases, a reference can be both a read _and_ a write. For
+        // example, `x += 1` is both a read of and a write to `x`.
+        edits.sort();
+        edits.dedup();
+
+        let edit = edits
+            .pop()
+            .ok_or(anyhow!("Unable to rename any references to `{name}`"))?;
+        Ok((edit, edits))
+    }
+
+    /// Rename a [`Binding`] reference.
+    fn rename_binding(binding: &Binding, target: &str) -> Option<Edit> {
+        match &binding.kind {
+            BindingKind::Importation(_) | BindingKind::FromImportation(_) => {
+                if binding.is_alias() {
+                    // Ex) Rename `import pandas as alias` to `import pandas as pd`.
+                    Some(Edit::range_replacement(target.to_string(), binding.range))
+                } else {
+                    // Ex) Rename `import pandas` to `import pandas as pd`.
+                    Some(Edit::insertion(
```

Comment:
do we need a pseudo edit to preserve `import pandas` here, too?

Reply:
Yes, good call.